compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -599,6 +601,26 @@ object Objects:
 
           case _ => a
 
+    def filterType(tpe: Type)(using Context): Value =
+      val baseClasses = tpe.baseClasses
+      if baseClasses.isEmpty then a
+      else tpe match
+        case t @ SAMType(_, _) if a.isInstanceOf[Fun] => a // if tpe is SAMType and a is Fun, allow it
```

Comment:
This case doesn't seem to use the baseclasses. Should all the baseclasses stuff (and the check for empty baseclasses) be pushed down into the second case? Also, perhaps `headOption` is useful on the baseclasses.